src/core/public/locale_helper.ts

```diff
@@ -0,0 +1,114 @@
+    }
+  }
+
+  return { localeValue: 'en', localeParam: null, updatedUrl: url };
+}
+
+/**
+ * Extracts a dynamically added locale parameter from a URL and restructures the URL
+ * to include this locale parameter in a consistent, functional manner.
+ *
+ * This function is specifically designed to handle cases where '&i18n-locale=<locale>'
+ * has been appended to the URL, potentially in a position that could cause issues
+ * with OpenSearch Dashboards' URL parsing or functionality.
+ *
+ * The restructuring is necessary because simply appending the locale parameter
+ * to certain URL structures can lead to parsing errors or the parameter being ignored.
+ *
+ * The function handles various URL structures to ensure the locale parameter
+ * is placed in a position where it will be correctly parsed and utilized by
+ * OpenSearch Dashboards, while maintaining the integrity of existing URL components.
+ *
+ * URL Components:
+ * - basePath: The part of the URL before the hash (#). It typically includes the domain and application path.
+ * - hashPath: The part of the URL after the hash (#) but before any query parameters (?).
+ * - hashQuery: The query parameters after the hashPath. In OpenSearch Dashboards, this often contains
+ *              RISON-encoded data and never ends with a slash (/) to avoid RISON parsing errors.
+ *
+ * This function handles three main categories of URLs:
+ * 1. basePath + # + hashPath (including when hashPath is '/' or empty)
+ *    Before: basePath#hashPath&i18n-locale=zh-CN
+ *    After:  basePath#hashPath?i18n-locale=zh-CN
+ *    Restructuring rationale: The '&' is changed to '?' because there were no existing
+ *    query parameters after the hashPath. This ensures the locale is treated as a proper
+ *    query parameter and not mistakenly considered part of the hashPath.
+ *
+ * 2. basePath + # + hashPath + ? + hashQuery
+ *    Before: basePath#hashPath?hashQuery&i18n-locale=zh-CN
+ *    After:  basePath#hashPath?hashQuery&i18n-locale=zh-CN
+ *    Restructuring rationale: The locale parameter is appended to existing query parameters.
+ *    No change in structure is needed as it's already in the correct position.
+ *
+ * 3. basePath only
+ *    Before: basePath&i18n-locale=zh-CN
+ *    After:  basePath?i18n-locale=zh-CN
+ *    Restructuring rationale: The '&' is changed to '?' because there were no existing
+ *    query parameters in the basePath. This ensures the locale is recognized as the
+ *    start of the query string rather than being misinterpreted as part of the path.
+ *
+ * The function performs the following steps:
+ * 1. Extracts the locale parameter from its current position in the URL.
+ * 2. Removes the locale parameter from its original position.
+ * 3. Reconstructs the URL, placing the locale parameter in the correct position
+ *    based on the URL structure to ensure proper parsing by OpenSearch Dashboards.
+ * 4. Updates the browser's URL without causing a page reload.
```

Comment:
skimmed through seems the purpose is to handle invalid URL param syntax where `&` should be used instead of `?`. Do we need to handle that? i'm not sure if anything else in OSD does similar handling, since usually it's user's responsibility to correctly use URL parameters

Reply:
From my tests, I found 
1) http://localhost:5601/app/home#/&i18n-locale=zh-CN will not work
2) http://localhost:5601/app/home#/? query= (...) &i18n-locale=zh-CN can work.

The error will cause a blank page without proper prompt error and hard to trace. I think /&i18n-locale=zh-CN might be interpreted as a path rather than a query parameter, leading to routing errors or the parameter being ignored.

Reply:
`http://localhost:5601/app/home#/&i18n-locale=zh-CN` this is not correct syntax, anything other parameters like `http://localhost:5601/app/home#/&xxx` also doesn't work. My point is there are many different support url params, do we need to do this special handling just for `i18n-locale`?

Reply:
the reason i'm asking this is that if user puts something like `/app/home#/&i18n-locale=en&embed=true` or the correct way `/app/home#/?i18n-locale=en&embed=true` i don't think this logic handles it, the implementation might not be generic enough to handle URL params

Reply:
Thank you for this feedback. I'd like to clarify a few points:

1) I assumed that `&i18n-locale` would always be attached at the end of the URL. Is this assumption correct in all cases?
2) Regarding format and validation, what specific aspects should we be concerned about? Are there any edge cases we need to handle?
3) For cases like `http://localhost:5601/app/home#/&i18n-locale=zh-CN`, should our function handle this, or should the app be responsible for ensuring the correct URL structure (ex http://localhost:5601/app/home/&i18n-locale=zh-CN) before our function is called?
4. If the app is responsible for URL structure, how should we document this requirement for developers using this feature?

Reply:
> 1. I assumed that &i18n-locale would always be attached at the end of the URL. Is this assumption correct in all cases?

I think we just want to give a reasonable API for client to override OSD locale, which URL parameter should be enough in this case. When this feature exposes to users we should advertise it as "adding a URL parameter of `i18n-locale` to change locale", not "append `&i18n-locale` to change locale", so i don't think we need to handle invalid URL parameters

> 2. Regarding format and validation, what specific aspects should we be concerned about? Are there any edge cases we need to handle?

Since there are similar existing implementation in OSD (`embed` parameter), I think following the same implementation should be enough

for 3 and 4 i'm not sure what "app" refers to, could you clarify? But user/consumer of the feature should be responsible to provide correct URL parameter, the function should read it in a way respecting to searchParams standard